src/storage/src/source/persist_source.rs

```diff
@@ -190,3 +194,218 @@ where
+    // operator in materialize/timely.
+
+    // This is a generator that sets up an async `Stream` that can be continously polled to get the
+    // values that are `yield`-ed from it's body.
+    let async_stream = async_stream::try_stream!({
+        // Worker 0 is responsible for distributing splits from listen
+        if worker_index != 0 {
+            trace!("We are not worker 0, exiting...");
+            return;
+        }
+
+        let persist_client = persist_clients_moveable
+            .lock()
+            .await
+            .open(persist_location_moveable)
+            .await
+            .expect("could not open persist client")
+            .clone();
+
+        let mut read = persist_client
+            .open_reader::<SourceData, (), mz_repr::Timestamp, mz_repr::Diff>(data_shard.clone())
+            .await
+            .expect("could not open persist shard");
+
+        // Aggressively downgrade `since`, to not hold back compaction.
+        read.downgrade_since(as_of_moveable.clone()).await;
+
+        let snapshots_hollow_batches = read
+            .snapshot_hollow_batch(as_of_moveable.clone())
+            .await
+            .expect("cannot serve requested as_of");
+
+        // First, yield all the updates from the snapshot.
+        for batch in snapshots_hollow_batches {
+            yield batch;
+        }
+
+        // Then, listen continously and yield any new updates. This loop is expected to never
+        // finish.
+        let mut listen = read
+            .listen(as_of_moveable)
+            .await
+            .expect("cannot serve requested as_of");
+
+        loop {
+            yield listen.listen_hollow_batch().await;
+        }
+    });
+
+    let mut pinned_stream = Box::pin(async_stream);
+
+    let (inner, token) =
+        crate::source::util::source(scope, "persist_source".to_string(), move |info| {
+            let waker_activator = Arc::new(scope.sync_activator_for(&info.address[..]));
+            let waker = futures_util::task::waker(waker_activator);
+            let activator = scope.activator_for(&info.address[..]);
+
+            // There is a bit of a mismatch: listening on a ReadHandle will give us an Antichain<T>
+            // as a frontier in `Progress` messages while a timely source usually only has a single
+            // `Capability` that it can downgrade. We can work around that by using a
+            // `CapabilitySet`, which allows downgrading the contained capabilities to a frontier
+            // but `util::source` gives us both a vanilla `Capability` and a `CapabilitySet` that
+            // we have to keep downgrading because we cannot simply drop the one that we don't
+            // need.
+
+            let mut current_ts = 0;
+
+            move |cap, output| {
+                let mut context = Context::from_waker(&waker);
+                // Bound execution of operator to prevent a single operator from hogging
```

Comment:
this is no longer necessary since it's not emitting updates but just low volume batches. we'll definitely need something like this in the second operator though, so maybe this comment is just "move this to the other operator plz"